File: 10001_20000/7001_8000/17953.cpp
#include <bits/stdc++.h>

using namespace std;

int main(){
	ios::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);  
  int day, kind;

  cin >> day >> kind;
  int** desert = new int*[10];

  for (int i = 0; i < kind; i++){
    desert[i] = new int[day];
  }

  for (int i = 0; i < kind; i++){
    for (int j = 0; j < day; j++){
      cin >> desert[i][j];
    }
  }

  int dp[10][2] = {0};


  for (int i = 0; i < kind; i++){
    dp[i][0] = desert[i][0];
  }

  int idx = 0;

  for (int i = 1; i < day; i++){
    for (int j = 0; j < kind; j++){
      for (int k = 0; k < kind; k++){
        int val = dp[k][idx] + (j == k ? desert[j][i] / 2 : desert[j][i]);
        if (val > dp[j][idx ^ 1]){
          dp[j][idx ^ 1] = val;
        }
      }
    }
    idx ^= 1;
  }

  int ans = -1;

  for (int i = 0; i < kind; i++){
    if (ans < dp[i][idx]){
      ans = dp[i][idx];
    }
  }

  cout << ans << endl;

}

File: 10001_20000/7001_8000/17953.py
from sys import stdin


input = stdin.readline


day, kind = map(int, input().split())

desert = []

for _ in range(kind):
    desert.append(list(map(int, input().split())))


dp = [[desert[i][0], 0] for i in range(kind)] 

idx = 0

for i in range(1, day):
    
    for j in range(kind):

        for k in range(kind):
            val = dp[k][idx] + (desert[j][i] if j != k else desert[j][i] // 2)
            # print('%d + %d = %d'%(dp[k][idx], (desert[j][i] if j != k else desert[j][i] // 2), val))
            if val > dp[j][idx ^ 1]:
                dp[j][idx ^ 1] = val

    idx ^= 1


ans = -1
for i in range(kind):
    if ans < dp[i][idx]:
        ans = dp[i][idx]

print(ans)
